src/tasks/withdrawService.ts

```diff
@@ -0,0 +1,385 @@
+   * retry.
+   */
+  retries: number;
+}
+
+function isValidState(state: unknown): state is State {
+  if (state === null || typeof state !== "object") {
+    console.error("State json is not an object");
+    return false;
+  }
+  const stateObject = state as State;
+  if (typeof stateObject["lastUpdateBlock"] !== "number") {
+    console.error("Invalid lastUpdateBlock");
+    return false;
+  }
+  if (typeof stateObject["nextTokenToTrade"] !== "number") {
+    console.error("Invalid nextTokenToTrade");
+    return false;
+  }
+  if (
+    !(
+      stateObject["tradedTokens"] instanceof Array &&
+      stateObject["tradedTokens"].every(
+        (elt) => typeof elt === "string" && utils.isAddress(elt),
+      )
+    )
+  ) {
+    console.error("Invalid tradedTokens");
+    return false;
+  }
+  if (
+    !(
+      stateObject["pendingTokens"] instanceof Array &&
+      stateObject["pendingTokens"].every(
+        (elt) =>
+          elt !== null &&
+          typeof elt === "object" &&
+          typeof elt.address === "string" &&
+          utils.isAddress(elt.address) &&
+          typeof elt.retries === "number",
+      )
+    )
+  ) {
+    console.error("Invalid pendingTokens");
+    return false;
+  }
+  return true;
+}
+
+interface WithdrawAndDumpInput {
+  state: State;
+  solver: SignerWithAddress;
+  receiver: string;
+  authenticator: Contract;
+  settlement: Contract;
+  settlementDeploymentBlock: number;
+  latestBlock: number;
+  minValue: string;
+  leftover: string;
+  validity: number;
+  maxFeePercent: number;
+  toToken: string;
+  network: SupportedNetwork;
+  usdReference: ReferenceToken;
+  hre: HardhatRuntimeEnvironment;
+  api: Api;
+  dryRun: boolean;
+}
+export async function withdrawAndDump({
+  state,
+  solver,
+  receiver,
+  authenticator,
+  settlement,
+  settlementDeploymentBlock,
+  latestBlock,
+  minValue,
+  leftover,
+  validity,
+  maxFeePercent,
+  toToken,
+  network,
+  usdReference,
+  hre,
+  api,
+  dryRun,
+}: WithdrawAndDumpInput): Promise<State> {
+  // Update list of pending tokens to determine which token was traded
+  const pendingTokens = (
+    await Promise.all(
+      state.pendingTokens.map(async (pendingToken) => {
+        if (pendingToken.retries >= MAX_ORDER_RETRIES_BEFORE_SKIPPING) {
+          // Note that this error might be triggered in legitimate cases, for
+          // example if a token did not trade the first time and then the price
+          // has become so low that it's not enough to pay for the fee.
+          // TODO: revisit after getting an idea of the frequency at which this
+          // alert is triggered.
+          console.error(
+            `Tried ${pendingToken.retries} times to sell token ${pendingToken.address} without success. Skipping token until future run`,
+          );
+          return [];
+        }
+
+        // assumption: eth is not in the list (as it's not supported by the
```

Comment:
nit: for a simple assumption like that maybe add a simple assert? Or does erc20Token throw if called on the native token?

Reply:
Done! It does happen that `erc20Token` reverts in this case, but I think we shouldn't be relying on this specific behavior of `erc20Token` since it might be changing in the future.